src/test/librados/asio.cc

```diff
@@ -83,7 +83,7 @@ TEST_F(AsioRados, AsyncReadCallback)
   librados::async_read(service, io, "exist", 256, 0, success_cb);
 
   auto failure_cb = [&] (boost::system::error_code ec, bufferlist bl) {
-    EXPECT_EQ(boost::system::errc::no_such_file_or_directory, ec);
+    EXPECT_EQ(boost::system::errc::no_such_file_or_directory, ec.value());
```

Comment:
I don't follow.  I believe this is exactly how `error_code` values should be compared since generic (`errc`) values are guaranteed to be a subset of system values.

What does `b` get assigned to in the following?
```
  bool b = (boost::system::errc::no_such_file_or_directory ==
            boost::system::error_code(ERROR_FILE_NOT_FOUND,
                                      boost::system::system_category()));
```

Reply:
Assuming the following assertions:

```
  EXPECT_EQ(boost::system::errc::no_such_file_or_directory,
            boost::system::error_code(ERROR_FILE_NOT_FOUND,
                                      boost::system::system_category()));

  EXPECT_EQ(boost::system::errc::read_only_file_system,
            boost::system::error_code(30,
                                      boost::system::system_category()));

  EXPECT_EQ(boost::system::errc::read_only_file_system,
            boost::system::error_code(30,
                                      boost::system::system_category()).value());
```

The second assertion fails, the others are passing.

```
Expected equality of these values:
  boost::system::errc::read_only_file_system
    Which is: 30
  boost::system::error_code(30, boost::system::system_category())
    Which is: system:30
```

It might be because ERROR_FILE_NOT_FOUND (2) happens to match ENOENT. On the other hand, we don't have an exact match for EROFS (30), ERROR_ACCESS_DENIED (5) would be the closest Windows equivalent.

Reply:
I've updated ``src/librados/librados_asio.h`` to use ``boost::system::generic_category`` on Windows.

Reply:
i think that's a good compromise, and follows the general guidance of https://www.boost.org/doc/libs/1_79_0/libs/system/doc/html/system.html#usage_returning_specific_errors_under_windows

it's not perfect, because the error codes from librados aren't guaranteed to be in the "portable subset" of errno values, but i agree this is the best short-term solution

cc @adamemerson who did a lot of work on error_codes for neorados

Reply:
> our own librados error category that would map 1:1 to Linux errno space

is that mapping a strict requirement? for librados, i agree that we do need to map everything to integer error codes. but maybe librados can add its own #defines for any error codes that aren't in this "portable subset" and map to those instead?

Reply:
Hrm, and it appears that Boost and STL are inconsistent even on Linux.  For POSIX errors, `system_category().default_error_condition(errno)` goes back to generic category on both (correct).  For non-POSIX errors, only STL sticks to system category:
```
- boost::system::system_category().default_error_condition(ENOENT) -> `generic:2`
- std::system_category().default_error_condition(ENOENT) -> `generic:2`
- boost::system::system_category().default_error_condition(ESHUTDOWN) -> `generic:108`   (!)
- std::system_category().default_error_condition(ENOENT) -> `system:108`
```

Reply:
> but maybe librados can add its own #defines for any error codes that aren't in this "portable subset" and map to those instead?

Yeah, I suppose theoretically that is an option but, given that there is an existing C/C++ API that would continue returning errors from the Linux errno space, having another set of #defines serve as an integer base for the new Boost ASIO API seems unnecessarily confusing.

Reply:
> having another set of #defines serving just as an integer base for the relatively new Boost ASIO API seems unnecessarily confusing to me

afaik, rgw is the only consumer of librados_asio.h, and that will eventually be superceded by neorados

i was just assuming that portable error codes were a problem in general. what does a windows client currently do with error codes like ESHUTDOWN and EKEYEXPIRED? i guess neorados is already presenting them via generic_category?

Reply:
Or perhaps, to avoid pitfalls like the "handle being blocklisted" example above, we could just unconditionally use (abuse) generic category on all platforms and document that for handling non-POSIX errors such as `ESHUTDOWN`, resorting to `ec.value() == FOO` tests is expected?  For POSIX errors, normally `ec == boost::system::errc::foo` tests would be used, of course.

I can see this being a not-so-short-term solution ;)

Reply:
> i guess neorados is already presenting them via generic_category?

neorados seems to be using `system_category`, hence my above proposal to switch to `generic_category`.

Reply:
> neorados seems to be using system_category, hence my above proposal to switch to generic_category.

i think it's more complicated than that. neorados already has its own [error_category](https://github.com/ceph/ceph/blob/9027acf/src/neorados/RADOS.cc#L1643), which inherits from [ceph::converting_category](https://github.com/ceph/ceph/blob/9027acf/src/common/error_code.h#L33). there's also a [ceph_category()](https://github.com/ceph/ceph/blob/9027acf/src/common/error_code.h#L38-L49) with its own enum of error conditions; maybe that's the best place to add the conditions for things like ESHUTDOWN and EKEYEXPIRED?

Reply:
Yeah, upon a closer look, neorados uses `system_category` in just a couple of places, for `EINVAL` and `ENOENT` (my statement was based on a context-less grep...).  `EKEYEXPIRED` was just an example, I don't think RADOS does anything with it.  `ESHUTDOWN` does get returned by OSDs though.

Reply:
@idryomov please let me know how we should proceed with this change. Should we stick with the current approach, ``generic_category`` on Windows and ``system_category`` on other platforms?

Reply:
I think we are waiting for input from @adamemerson.  That said, since this change is a change only for Windows, it's should be safe to merge.